pkg/controller/node/node_controller.go

```diff
@@ -128,6 +143,199 @@ func (ctrl *Controller) Run(workers int, stopCh <-chan struct{}) {
+			// Remove master taint
+			newTaints := []corev1.Taint{}
+			for _, t := range node.Spec.Taints {
+				if t.Key == "node-role.kubernetes.io/master" && t.Effect == corev1.TaintEffectNoSchedule {
+					continue
+				}
+				newTaints = append(newTaints, t)
+			}
+			//
+			node.Spec.Taints = newTaints
+		})
+		if err != nil {
+			glog.Errorf("error making node %v schedulable with error %v", err)
+		}
+	}
+}
+
+// makeMasterNodesUnSchedulable takes currentSchedulabeMasters and the number of master nodes which need to be made
+// schedulable as requested by user. As of now, we take schedulable master nodes slice and iterate over it to make
+// them schedulable.
+func (crtl *Controller) makeMasterNodesUnSchedulable(currentSchedulableMasters []*corev1.Node, neededUnSchedulableMasters int) {
+	for i := 0; i < neededUnSchedulableMasters; i++ {
+		_, err := updateNodeRetry(crtl.kubeClient.CoreV1().Nodes(), crtl.nodeLister, currentSchedulableMasters[i].Name, func(node *corev1.Node) {
+			// Remove worker label
+			newLabels := node.Labels
+			if _, hasWorkerLabel := newLabels["node-role.kubernetes.io/worker"]; hasWorkerLabel {
+				delete(newLabels, "node-role.kubernetes.io/worker")
+			}
+			node.Labels = newLabels
+			// Add master taint
+			newTaints := node.Spec.Taints
+			MasterUnSchedulableTaint := corev1.Taint{Key: "node-role.kubernetes.io/master", Effect: corev1.TaintEffectNoSchedule}
+			newTaints = append(newTaints, MasterUnSchedulableTaint)
+			node.Spec.Taints = newTaints
+		})
+		if err != nil {
+			glog.Errorf("error making node %v unschedulable with error %v", err)
+		}
+	}
+}
+
+// This is duplicated code and is copied from daemon. Perhaps a more neutral place for this exists. If not can we
+// create one?
+func updateNodeRetry(client corev1client.NodeInterface, lister corev1lister.NodeLister, nodeName string, f func(*corev1.Node)) (*corev1.Node, error) {
+	var node *corev1.Node
+	if err := retry.RetryOnConflict(retry.DefaultBackoff, func() error {
+		n, err := lister.Get(nodeName)
+		if err != nil {
+			return err
+		}
+		oldNode, err := json.Marshal(n)
+		if err != nil {
+			return err
+		}
+
+		nodeClone := n.DeepCopy()
+		f(nodeClone)
+
+		newNode, err := json.Marshal(nodeClone)
+		if err != nil {
+			return err
+		}
+
+		patchBytes, err := strategicpatch.CreateTwoWayMergePatch(oldNode, newNode, corev1.Node{})
+		if err != nil {
+			return fmt.Errorf("failed to create patch for node %q: %v", nodeName, err)
+		}
+
+		node, err = client.Patch(nodeName, types.StrategicMergePatchType, patchBytes)
+		return err
+	}); err != nil {
+		// may be conflict if max retries were hit
+		return nil, fmt.Errorf("unable to update node %q: %v", node, err)
+	}
+	return node, nil
+}
+
+func (crtl *Controller) getCurrentMasters() []*corev1.Node {
+	nodeList, err := crtl.nodeLister.List(labels.SelectorFromSet(labels.Set{"node-role.kubernetes.io/master": ""}))
+	if err != nil {
+		glog.Errorf("error while listing master nodes %v", err)
```

Comment:
This seems like a particularly egregious "log and continue" case.  While it's *probably* safe to have the rest of the code operate on the empty array in an error situation, why even worry about it?  Let's error out early and avoid thinking about it.

Reply:
As I told earlier, I have seen both the patterns. Either log and continue or return early with an error. I agree that in this case, perhaps error-ing early might help. As I told this is rough shape of what you can see, not the exact fn.